File: triangle.py
class Solution:
    # @param triangle, a list of lists of integers
    # @return an integer
    def minimumTotal(self, triangle):
        l = len(triangle) - 2
        while l >= 0:
        	for j in range(len(triangle[l])):
        		triangle[l][j] += min(triangle[l+1][j], triangle[l+1][j+1])
        	l -= 1
        return triangle[0][0]

s = Solution()
triangle = [\
     [2],\
  #   [3,4],\
  #  [6,5,7],\
  # [4,1,8,3]\
]
print s.minimumTotal(triangle)

File: JAVA/triangle.java
public class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        ArrayList<Integer> lastRow = (ArrayList<Integer>) triangle.get(triangle.size()-1);
        ArrayList<Integer> currentRow = new ArrayList<Integer>();
        int count = triangle.size()-2;
        while (count >= 0){
            for (int i = 1; i < lastRow.size(); i++)
                currentRow.add(triangle.get(count).get(i-1) + Math.min(lastRow.get(i-1),lastRow.get(i)));
            count--;
            lastRow = currentRow;
            currentRow = new ArrayList<Integer>();
        }
        return lastRow.get(0);
    }
}